### DISCUSSION TITLE
How do you handle page specific components inside static layouts?

### DISCUSSION THREAD

**Discussion by michaelschufi** [vercel/next.js#54154] [App Router]
Suppose you have a root layout with a header, footer, and main. The header includes the whole navigation logic and the footer secondary navigations and other sections. The page is rendered inside main, while the header and the footer are mostly static.

Now, what do I mean by mostly... A few parts are dependent on the current route or page data.

- The currently active navigation item should be marked as active.
- A language switcher component (potentially deeply nested in the header/footer) needs links for other versions of the same page (e.g. translated URL segments).

The first case should be trivial, because you can just use `usePathname` and match it with the navigation item's `href`.

However, the second case needs data from the backend (the translated URLs of the current page), which is not straightforward.
Can you somehow SSR and cache the language switcher? How would you do this without having to do a client side fetching? (Client side fetching is worse UX than if it's just a server component that is loaded based on the page data you already fetch for each page.)

Edit:
Could you maybe somehow pass data or a component from the page to the layout? E.g. similar to the page being passed as "children" to the layout.

This is the biggest pain point for us right now.

Don't get me wrong. The layouts, nested layouts, route group features are all so awesome, but if you want to have a single dynamic part in your layout, I *think* you need to fall back to a client component right now, even though the page is SSRing anyway. Is this true, or is there an easy workaround that I have completely missed?

**Comment by tylerpaige** [comment]
@michaelschufi I'm curious if you've figured out a solution for this problem, as I am struggling with the same thing.

In my case, I want to be able to display a title in a `<Header/>` component which is rendered in the layout. The title is set on a per-page basis. Therefore I would love to pass the title from the page to layout. 

I've worked out a imperfect solution using contexts and client-side rendering, but it has unfortunately has a flash in between the initial page load and the client-side rendering.


```jsx
// NOTE: I would not recommend copying this code :-) 

// header context
export const HeaderContext = createContext({
  title: '',
  setTitle: () => {},
});

export function HeaderProvider({ children, ...props }) {
  const [title, setTitle] = useState(false);

  return (
    <HeaderContext.Provider
      value={{
        title,
        setTitle
      }}
    >
      {children}
    </HeaderContext.Provider>
  );
}

// layout
export default async function RootLayout({ children }) {
  return (
    <html>
      <body>
        <HeaderProvider>
          <Header />
          { children }
          <Footer />
        </HeaderProvider>
      </body>
    </html>
  )
}

// header
'use client';
export function Header() {
  const { title } = useContext(HeaderContext);

  return (
    <header>
      <h1>{title}</h1>
    </header>
  )
}

// page
export default async function SomePage() {
  return (
    <div>
      <PageHeader title="Page Title" />
    </div>
  )
}

// page header
'use client'
export function PageHeader({ title }) {
  const { setTitle } = useContext(HeaderContext);
  
  useEffect(() => {
    setTitle(title);
  }, [title]);

  return (
    <header>
      <h1>{title}</h1>
    </header>
  )
}
```

**Reply by michaelschufi** [reply]
In short, no. Not a satisfactory solution at least - just workarounds.

A discussion in Discord revealed that it's not terrible for a page to have parts of the layout, because through good caching you should only experience a simple rerender while navigating. I still think this defeats the point of layouts, though. 🤷‍♂️
A layout, right now at least 🤞, is very static (only rerenders if a param changes, afaik).

Alternatively, you could simply make a client side request to get whatever data you need.  IMO, this is only fine if you don't need this data immediately, e.g. a hidden navigation, because I don't like loading states...

**Reply by tylerpaige** [reply]
agreed re layouts. i'm curious if this is a good usecase for templates maybe? at any rate, i think you're right that moving some components from layouts to pages is probably the way to go. thanks!

**Reply by landsman** [reply]
I just did the same workaround as you here. But I really don't like it 😢

**Comment by landsman** [comment]
This topic is relevant for me as well. It's a standard thing in different lframeworks (for example php - nette / symfony) when you are able to "rewrite" something like "snippet" from bottom to above (child -> parent).

**Reply by raphaelpc** [reply]
Worst part is that using Pages router that is totally doable:
https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts#per-page-layouts

To me this is a huge downside of the APP router =/

**Comment by ignaciogiri** [comment]
Hi! I’d still love to know the latest recommended approach (as of today) in the App Router to achieve this kind of persistent header / static shell behavior across route changes.

I tried route groups + per-layout headers, but the header unmounts/remounts between routes and causes a visible flickering.

What’s the recommended approach today to keep the shell mounted and just swap the page content?

Options I’m considering:

- `useSelectedLayoutSegment(s)` to drive header state
- Parallel routes for a persistent shell, slots? I'm lost
- A client HeaderProvider (state decoupled from layouts)

Any guidance or a reference implementation would be super helpful.

<img width="1920" height="1080" alt="1" src="https://github.com/user-attachments/assets/30def79b-938c-4a15-94e9-88f2c2528237" />

<img width="1920" height="1080" alt="2" src="https://github.com/user-attachments/assets/52329d8c-9aba-476c-a4b0-fb3b6e09ffbb" />
